### PR TITLE
[composable] Enable replicate + trec_shard overall

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98890
* #98889

replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

### CODE REVIEW

File: torch/distributed/_composable/replicate.py
Comment (mrshenli) [inline torch/distributed/_composable/replicate.py:43]: any reason for removing this?

We will remove this eventually, but curious about does this prevent integrating with trec_shard?

File: torch/distributed/_composable/replicate.py
Comment (mrshenli) [inline torch/distributed/_composable/replicate.py:56]: when, without this, will it still work if a submodule is wrapped by `fully_shard`?

File: torch/distributed/_composable/replicate.py
Comment (mrshenli) [inline torch/distributed/_composable/replicate.py:14]: curious, can trec_shard add entry to registry on submodules it swapped, so that replicate can detect automatically?

Comment (yhcharles) [reply torch/distributed/_composable/replicate.py:14]: Agree, and that's the spirit of these *composable* APIs, `ignore_modules` may look like a temporary workaround solution.

File: torch/distributed/_composable/replicate.py
Comment (yhcharles) [inline torch/distributed/_composable/replicate.py:56]: The original idea to make `replicate` to be composable is: assuming we use it on a given Module, then it will recursively check all the sub-modules, apply data parallel on the sub-modules only when the sub-module is not marked by another composable API (since all composable APIs follows `contract`, they are supposed to be able to know each other).
It seems this change removes this logic, and makes `replicate` not composable. Is there any reason or it's my misunderstanding?
cc @mrshenli

File: torch/distributed/_composable/replicate.py
Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:43]: Actually we can keep this, because _can_compose only checked whether replicate can wrap a fully_shard which is unsupported at the moment.

File: torch/distributed/_composable/replicate.py
Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:14]: This is indeed the right way to do so. currently trec_shard adds an entry for the top level, but does not do so for individually sharded submodules. 

This is also why fully_shard needs ignored_modules today. For this PR, I think it'd be best to unblock by adding ignored_module support, and then follow up with trec_shard to fix the registry and remove ignored_modules.

Also, if user calls replicate() first and then trec_shard, we would still need `ignored_modules`, right? Unless we add a policy or similar to replicate API.

File: torch/distributed/_composable/replicate.py
Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:56]: IIUC this was not completely the previous behavior.

Previously, `_can_compose` only checked for fully_shard, so trec_shard would get through. Thus when I tried to do a top-level replicate, replicate would try to mark embedding modules as data parallel which is not desired. 

The right fix would be to add the registry check for trec_shard as exists for _can_compose. However, trec_shard does not seem to add to the registry when it is wrapping an embedding submodule (only adds it for the top level). Once we fix this in trec_shard we may be able to get rid of ignore_modules. 

We should also make sure fully_shard can get rid of it as well then.

Comment (yhcharles) [reply torch/distributed/_composable/replicate.py]: Yes, the previous implementation was not complete -- only `fully_shard` and `replicate` run the check and thus are composable with each other. And also agree that the right fix is to enable the check for trec_shard.

Q:
- If we plan to add this check to trec_shard, then why removing it from `replicate`? It seems adding `ignored_modules` has not conflicts with the check.
- Will removing this check break the existing composability of `fully_shard` and `replicate`?

File: test/distributed/_composable/test_replicate.py
Comment (mrshenli) [inline test/distributed/_composable/test_replicate.py:136]: default world size in `MultiProcessTestCase` is 4. We will need to override `world_size` property or change this to 4 as well.

```python
@property
    def world_size(self) -> int:
      return 2
```

**Review by mrshenli — APPROVED** [review]
stamp to unblock, please fix lint before landing.

File: torch/distributed/_composable/replicate.py
Comment (yhcharles) [inline torch/distributed/_composable/replicate.py:77]: Thanks, LGTM

**Review by yhcharles — APPROVED** [review]

**Pull request closed without merging**
